File: src/lib/engine/location.ts
import { nearestSessionStructure, type SessionLevel } from "./sessions";
import type { SignalEvaluation, TradeDirection } from "./quant";
import type { BaseZone } from "./zones";

/**
 * Entry-location grading. A directionally-correct verdict is still a bad trade
 * when price is extended away from the structure you'd want to enter against —
 * a long chased into resistance has a far stop and poor reward:risk even though
 * the thesis is right. This grades *where* price sits in the support→resistance
 * range for the trade's direction, so "favored" can be reserved for setups that
 * are also well located, and everything else can say "right direction, wrong
 * price — wait for a pullback to $X".
 *
 * Grading runs on the last *closed* bar's analytics (same basis as the rest of
 * the engine), so it changes once per bar rather than flickering intra-bar.
 */

export type LocationGrade = "at-structure" | "mid-range" | "extended";

/** How much fresh supply/demand structure backs the entry location. */
export type ZoneConfluence = "none" | "single-timeframe" | "multi-timeframe";

export interface LocationRead {
  grade: LocationGrade;
  /** Short chip label for the UI. */
  label: string;
  /** One-sentence plain explanation. */
  note: string;
  /** The level to wait for a pullback/rally toward (support for a long, resistance for a short). */
  pullbackTarget: number;
  /** 0 = at support, 1 = at resistance. May sit slightly outside [0,1] when price breaks a level. */
  rangePosition: number;
  /**
   * Fresh supply/demand backing: "none", "single-timeframe" (price sits on a
   * zone on the execution timeframe), or "multi-timeframe" (a higher-timeframe
   * zone lines up too — the confluence discretionary traders wait for).
   */
  confluence: ZoneConfluence;
}

/** Execution- and context-timeframe zones for grading + confluence. */
export interface LocationZones {
  execution: BaseZone[];
  context: BaseZone[];
}

const LABELS: Record<LocationGrade, string> = {
  "at-structure": "At structure",
  "mid-range": "Mid-range",
  extended: "Extended",
};

function fmtPrice(value: number): string {
  const abs = Math.abs(value);
  if (abs >= 1000) return `$${value.toLocaleString("en-US", { maximumFractionDigits: 2 })}`;
  if (abs >= 1) return `$${value.toLocaleString("en-US", { maximumFractionDigits: 4 })}`;
  return `$${value.toLocaleString("en-US", { maximumSignificantDigits: 5 })}`;
}

function clamp(value: number, lo: number, hi: number): number {
  return Math.min(hi, Math.max(lo, value));
}

function zonesOverlap(a: BaseZone, b: BaseZone): boolean {
  return a.priceLow <= b.priceHigh && a.priceHigh >= b.priceLow;
}

/**
 * The best fresh/tested zone price is sitting on for the trade direction: a
 * demand zone at/just-below price for a long, a supply zone at/just-above for a
 * short. Fresh zones win over tested; ties break to the nearest proximal edge.
 */
function entryZoneAtPrice(
  zones: BaseZone[],
  direction: "long" | "short",
  price: number,
  atr: number | null,
): BaseZone | null {
  const kind = direction === "long" ? "demand" : "supply";
  const matches = zones.filter((z) => {
    if (z.kind !== kind) return false;
    const buffer = atr && atr > 0 ? atr * 0.5 : (z.priceHigh - z.priceLow) * 0.5 || price * 0.002;
    return direction === "long"
      ? price >= z.priceLow && price <= z.priceHigh + buffer
      : price <= z.priceHigh && price >= z.priceLow - buffer;
  });
  if (matches.length === 0) return null;
  return matches.sort((a, b) => {
    if (a.freshness !== b.freshness) return a.freshness === "fresh" ? -1 : 1;
    const pa = direction === "long" ? a.priceHigh : a.priceLow;
    const pb = direction === "long" ? b.priceHigh : b.priceLow;
    return Math.abs(price - pa) - Math.abs(price - pb);
  })[0];
}

/** The nearest fresh directional zone still ahead of a pullback (below for a long). */
function pullbackZone(
  zones: BaseZone[],
  direction: "long" | "short",
  price: number,
): BaseZone | null {
  const kind = direction === "long" ? "demand" : "supply";
  const candidates = zones.filter(
    (z) =>
      z.kind === kind &&
      z.freshness === "fresh" &&
      (direction === "long" ? z.priceHigh < price : z.priceLow > price),
  );
  if (candidates.length === 0) return null;
  return candidates.sort((a, b) =>
    direction === "long" ? b.priceHigh - a.priceHigh : a.priceLow - b.priceLow,
  )[0];
}

/**
 * Grades where price sits for the given direction. Returns null when there
 * isn't a usable support/resistance frame to judge against. When `zones` is
 * supplied, fresh supply/demand structure can promote the grade to
 * at-structure and flag multi-timeframe confluence.
 */
export function gradeLocation(
  evaluation: SignalEvaluation,
  direction: TradeDirection,
  zones?: LocationZones,
  sessionLevels?: SessionLevel[],
): LocationRead | null {
  if (direction !== "long" && direction !== "short") return null;
  const { support, resistance, lastClose, atrPercent } = evaluation.analytics;
  if (support === null || resistance === null || !lastClose || resistance <= support) return null;

  // Fraction of the support→resistance band where price sits.
  const pos = clamp((lastClose - support) / (resistance - support), -0.5, 1.5);
  // Proximity to the *ideal* entry structure: 0 = right at it, 1 = chasing into
  // the opposite level. Long enters near support; short near resistance.
  const proximity = direction === "long" ? pos : 1 - pos;
  const pullbackTarget = direction === "long" ? support : resistance;

  // In a wide range, price can be within a bar's worth of the level yet still
  // read mid-range by fraction alone — treat "within ~1 ATR of structure" as
  // at-structure regardless.
  const atrPct = atrPercent && atrPercent > 0 ? atrPercent : null;
  const distToStructurePct =
    direction === "long"
      ? ((lastClose - support) / lastClose) * 100
      : ((resistance - lastClose) / lastClose) * 100;
  const hugsStructure = atrPct !== null && distToStructurePct <= atrPct * 1.1;

  let grade: LocationGrade;
  if (proximity <= 0.4 || hugsStructure) grade = "at-structure";
  else if (proximity >= 0.7) grade = "extended";
  else grade = "mid-range";

  const nearLevel = direction === "long" ? support : resistance;
  const farLevel = direction === "long" ? resistance : support;
  const side = direction === "long" ? "support" : "resistance";
  const ceiling = direction === "long" ? "resistance" : "support";
  const zoneWord = direction === "long" ? "demand" : "supply";

  // Fresh supply/demand backing. A zone price is literally sitting on is a
  // better read of "at structure" than a pivot fraction, so it can promote the
  // grade; a higher-timeframe zone lining up is multi-timeframe confluence.
  const atr = evaluation.analytics.atr14;
  const exeZone = zones ? entryZoneAtPrice(zones.execution, direction, lastClose, atr) : null;
  let pullbackLevel = pullbackTarget;
  let confluence: ZoneConfluence = "none";
  let note: string;

  if (exeZone) {
    grade = "at-structure";
    const higherZone = zones?.context.find(
      (z) =>
        z.kind === exeZone.kind &&
        (zonesOverlap(z, exeZone) || (lastClose >= z.priceLow && lastClose <= z.priceHigh)),
    );
    confluence = higherZone ? "multi-timeframe" : "single-timeframe";
    note =
      confluence === "multi-timeframe"
        ? `Price is on a ${exeZone.freshness} ${zoneWord} zone (${fmtPrice(exeZone.priceLow)}–${fmtPrice(exeZone.priceHigh)}) that lines up with a higher-timeframe ${zoneWord} zone — multi-timeframe confluence, the highest-quality ${direction} entry on offer.`
        : `Price is on a ${exeZone.freshness} ${zoneWord} zone (${fmtPrice(exeZone.priceLow)}–${fmtPrice(exeZone.priceHigh)}) — a structure-backed ${direction} entry with a tight stop below it.`;
  } else if (grade === "at-structure") {
    note = `Price is hugging ${side} (${fmtPrice(nearLevel)}) — a clean spot to enter a ${direction} with a tight stop.`;
  } else if (grade === "mid-range") {
    note = `Price is mid-range between ${fmtPrice(support)} and ${fmtPrice(resistance)} — a workable but not ideal ${direction} entry; a pullback toward ${fmtPrice(pullbackTarget)} would be cleaner.`;
  } else {
    // When extended, a fresh zone beyond price is a better pullback target than
    // the raw pivot — that's where a discretionary trader would wait.
    const zone = zones ? pullbackZone(zones.execution, direction, lastClose) : null;
    if (zone) pullbackLevel = direction === "long" ? zone.priceHigh : zone.priceLow;
    note = zone
      ? `Price is extended toward ${ceiling} (${fmtPrice(farLevel)}). Wait for a pullback into the fresh ${zoneWord} zone near ${fmtPrice(pullbackLevel)} — a far cleaner ${direction} entry than chasing here.`
      : `Price is extended toward ${ceiling} (${fmtPrice(farLevel)}) — entering here chases the move with a wide stop and little room to target. Wait for a pullback toward ${fmtPrice(pullbackLevel)}.`;
  }

  // Session high/low levels are horizontal intraday structure. A session level
  // price is holding on the entry side is a concrete place to lean a stop —
  // strong enough to read as "at structure" on its own, and extra confluence
  // when it stacks on a supply/demand zone.
  const sessionHugDist = atr && atr > 0 ? atr * 0.75 : lastClose * 0.004;
  const sessionHug = sessionLevels?.length
    ? nearestSessionStructure(sessionLevels, direction, lastClose, sessionHugDist)
    : null;
  if (sessionHug) {
    const held = direction === "long" ? "holding" : "capped at";
    if (exeZone) {
      note += ` It stacks on the ${sessionHug.label} session ${sessionHug.kind} (${fmtPrice(sessionHug.price)}) — added session-level confluence.`;
    } else if (grade !== "at-structure") {
      grade = "at-structure";
      note = `Price is ${held} the ${sessionHug.label} session ${sessionHug.kind} (${fmtPrice(sessionHug.price)}) — an intraday level to lean a tight stop against for a ${direction} entry.`;
    } else {
      note += ` It also aligns with the ${sessionHug.label} session ${sessionHug.kind} (${fmtPrice(sessionHug.price)}).`;
    }
  }

  return {
    grade,
    label: LABELS[grade],
    note,
    pullbackTarget: pullbackLevel,
    rangePosition: pos,
    confluence,
  };
}

File: engine/smc/location.py
"""Entry-location grading (port of location.ts).

A directionally-correct verdict is still a bad trade when price is extended
away from the structure you'd want to enter against. This grades *where*
price sits in the support->resistance range for the trade's direction, so
"favored" can be reserved for setups that are also well located.

Grading runs on the last *closed* bar's analytics, so it changes once per bar.
"""

from dataclasses import dataclass
from typing import Literal

from smc.quant import SignalEvaluation, TradeDirection
from smc.sessions import SessionLevel, nearest_session_structure
from smc.zones import BaseZone

LocationGrade = Literal["at-structure", "mid-range", "extended"]
# How much fresh supply/demand structure backs the entry location.
ZoneConfluence = Literal["none", "single-timeframe", "multi-timeframe"]

_LABELS: dict[LocationGrade, str] = {
    "at-structure": "At structure",
    "mid-range": "Mid-range",
    "extended": "Extended",
}


@dataclass(slots=True)
class LocationRead:
    grade: LocationGrade
    # Short chip label for the UI.
    label: str
    # One-sentence plain explanation.
    note: str
    # The level to wait for a pullback/rally toward.
    pullback_target: float
    # 0 = at support, 1 = at resistance. May sit slightly outside [0,1].
    range_position: float
    confluence: ZoneConfluence


@dataclass(slots=True)
class LocationZones:
    """Execution- and context-timeframe zones for grading + confluence."""

    execution: list[BaseZone]
    context: list[BaseZone]


def _fmt_price(value: float) -> str:
    abs_ = abs(value)
    if abs_ >= 1000:
        text = f"{value:,.2f}".rstrip("0").rstrip(".")
        return f"${text}"
    if abs_ >= 1:
        text = f"{value:,.4f}".rstrip("0").rstrip(".")
        return f"${text}"
    text = f"{value:.5g}"
    return f"${text}"


def _clamp(value: float, lo: float, hi: float) -> float:
    return min(hi, max(lo, value))


def _zones_overlap(a: BaseZone, b: BaseZone) -> bool:
    return a.price_low <= b.price_high and a.price_high >= b.price_low


def _entry_zone_at_price(
    zones: list[BaseZone],
    direction: Literal["long", "short"],
    price: float,
    atr: float | None,
) -> BaseZone | None:
    """The best fresh/tested zone price is sitting on for the trade direction.
    Fresh zones win over tested; ties break to the nearest proximal edge."""
    kind = "demand" if direction == "long" else "supply"
    matches: list[BaseZone] = []
    for z in zones:
        if z.kind != kind:
            continue
        buffer = (
            atr * 0.5
            if atr is not None and atr > 0
            else ((z.price_high - z.price_low) * 0.5 or price * 0.002)
        )
        inside = (
            z.price_low <= price <= z.price_high + buffer
            if direction == "long"
            else z.price_low - buffer <= price <= z.price_high
        )
        if inside:
            matches.append(z)
    if not matches:
        return None
    return sorted(
        matches,
        key=lambda z: (
            0 if z.freshness == "fresh" else 1,
            abs(price - (z.price_high if direction == "long" else z.price_low)),
        ),
    )[0]


def _pullback_zone(
    zones: list[BaseZone],
    direction: Literal["long", "short"],
    price: float,
) -> BaseZone | None:
    """The nearest fresh directional zone still ahead of a pullback (below for a long)."""
    kind = "demand" if direction == "long" else "supply"
    candidates = [
        z
        for z in zones
        if z.kind == kind
        and z.freshness == "fresh"
        and (z.price_high < price if direction == "long" else z.price_low > price)
    ]
    if not candidates:
        return None
    return sorted(
        candidates,
        key=lambda z: -z.price_high if direction == "long" else z.price_low,
    )[0]


def grade_location(
    evaluation: SignalEvaluation,
    direction: TradeDirection,
    zones: LocationZones | None = None,
    session_levels: list[SessionLevel] | None = None,
) -> LocationRead | None:
    """Grades where price sits for the given direction. None when there isn't a
    usable support/resistance frame. Fresh supply/demand structure can promote
    the grade to at-structure and flag multi-timeframe confluence."""
    if direction not in ("long", "short"):
        return None
    analytics = evaluation.analytics
    support = analytics.support
    resistance = analytics.resistance
    last_close = analytics.last_close
    atr_percent = analytics.atr_percent
    if support is None or resistance is None or not last_close or resistance <= support:
        return None

    # Fraction of the support->resistance band where price sits.
    pos = _clamp((last_close - support) / (resistance - support), -0.5, 1.5)
    # Proximity to the *ideal* entry structure: 0 = right at it, 1 = chasing
    # into the opposite level.
    proximity = pos if direction == "long" else 1 - pos
    pullback_target = support if direction == "long" else resistance

    # "Within ~1 ATR of structure" reads as at-structure regardless of fraction.
    atr_pct = atr_percent if atr_percent is not None and atr_percent > 0 else None
    dist_to_structure_pct = (
        ((last_close - support) / last_close) * 100
        if direction == "long"
        else ((resistance - last_close) / last_close) * 100
    )
    hugs_structure = atr_pct is not None and dist_to_structure_pct <= atr_pct * 1.1

    grade: LocationGrade
    if proximity <= 0.4 or hugs_structure:
        grade = "at-structure"
    elif proximity >= 0.7:
        grade = "extended"
    else:
        grade = "mid-range"

    near_level = support if direction == "long" else resistance
    far_level = resistance if direction == "long" else support
    side = "support" if direction == "long" else "resistance"
    ceiling = "resistance" if direction == "long" else "support"
    zone_word = "demand" if direction == "long" else "supply"

    # Fresh supply/demand backing. A zone price is literally sitting on is a
    # better read of "at structure" than a pivot fraction, so it can promote
    # the grade; a higher-timeframe zone lining up is multi-timeframe confluence.
    atr = analytics.atr14
    exe_zone = _entry_zone_at_price(zones.execution, direction, last_close, atr) if zones else None
    pullback_level = pullback_target
    confluence: ZoneConfluence = "none"

    if exe_zone is not None:
        grade = "at-structure"
        higher_zone = (
            next(
                (
                    z
                    for z in zones.context
                    if z.kind == exe_zone.kind
                    and (_zones_overlap(z, exe_zone) or z.price_low <= last_close <= z.price_high)
                ),
                None,
            )
            if zones
            else None
        )
        confluence = "multi-timeframe" if higher_zone is not None else "single-timeframe"
        zone_span = f"{_fmt_price(exe_zone.price_low)}–{_fmt_price(exe_zone.price_high)}"  # noqa: RUF001 — en dash matches the TS copy
        note = (
            (
                f"Price is on a {exe_zone.freshness} {zone_word} zone ({zone_span}) that lines "
                f"up with a higher-timeframe {zone_word} zone — multi-timeframe confluence, "
                f"the highest-quality {direction} entry on offer."
            )
            if confluence == "multi-timeframe"
            else (
                f"Price is on a {exe_zone.freshness} {zone_word} zone ({zone_span}) — a "
                f"structure-backed {direction} entry with a tight stop below it."
            )
        )
    elif grade == "at-structure":
        note = (
            f"Price is hugging {side} ({_fmt_price(near_level)}) — a clean spot to enter a "
            f"{direction} with a tight stop."
        )
    elif grade == "mid-range":
        note = (
            f"Price is mid-range between {_fmt_price(support)} and {_fmt_price(resistance)} — "
            f"a workable but not ideal {direction} entry; a pullback toward "
            f"{_fmt_price(pullback_target)} would be cleaner."
        )
    else:
        # When extended, a fresh zone beyond price is a better pullback target
        # than the raw pivot — that's where a discretionary trader would wait.
        zone = _pullback_zone(zones.execution, direction, last_close) if zones else None
        if zone is not None:
            pullback_level = zone.price_high if direction == "long" else zone.price_low
        note = (
            (
                f"Price is extended toward {ceiling} ({_fmt_price(far_level)}). Wait for a "
                f"pullback into the fresh {zone_word} zone near {_fmt_price(pullback_level)} — "
                f"a far cleaner {direction} entry than chasing here."
            )
            if zone is not None
            else (
                f"Price is extended toward {ceiling} ({_fmt_price(far_level)}) — entering here "
                f"chases the move with a wide stop and little room to target. Wait for a "
                f"pullback toward {_fmt_price(pullback_level)}."
            )
        )

    # Session high/low levels are horizontal intraday structure — strong enough
    # to read as "at structure" on their own, extra confluence when stacked.
    session_hug_dist = atr * 0.75 if atr is not None and atr > 0 else last_close * 0.004
    session_hug = (
        nearest_session_structure(session_levels, direction, last_close, session_hug_dist)
        if session_levels
        else None
    )
    if session_hug is not None:
        held = "holding" if direction == "long" else "capped at"
        session_desc = (
            f"the {session_hug.label} session {session_hug.kind} ({_fmt_price(session_hug.price)})"
        )
        if exe_zone is not None:
            note += f" It stacks on {session_desc} — added session-level confluence."
        elif grade != "at-structure":
            grade = "at-structure"
            note = (
                f"Price is {held} {session_desc} — an intraday level to lean a tight stop "
                f"against for a {direction} entry."
            )
        else:
            note += f" It also aligns with {session_desc}."

    return LocationRead(
        grade=grade,
        label=_LABELS[grade],
        note=note,
        pullback_target=pullback_level,
        range_position=pos,
        confluence=confluence,
    )
